hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ExitCallsEnricher.java

```diff
@@ -0,0 +1,116 @@
+import org.hypertrace.core.datamodel.shared.ApiNode;
+import org.hypertrace.core.datamodel.shared.trace.AttributeValueCreator;
+import org.hypertrace.traceenricher.enrichedspan.constants.EnrichedSpanConstants;
+import org.hypertrace.traceenricher.enrichedspan.constants.utils.EnrichedSpanUtils;
+import org.hypertrace.traceenricher.enrichment.AbstractTraceEnricher;
+import org.hypertrace.traceenricher.trace.util.ApiTraceGraph;
+
+public class ExitCallsEnricher extends AbstractTraceEnricher {
+
+  @Override
+  public void enrichTrace(StructuredTrace trace) {
+    Map<ByteBuffer, ApiExitCallInfo> exitCallInfo = computeApiExitCallCount(trace);
+    trace.getEventList().stream()
+        .filter(event -> exitCallInfo.containsKey(event.getEventId()))
+        .forEach(
+            event -> {
+              ApiExitCallInfo apiExitCallInfo = exitCallInfo.get(event.getEventId());
+              addEnrichedAttribute(
+                  event,
+                  EnrichedSpanConstants.API_EXIT_CALLS_ATTRIBUTE,
+                  AttributeValueCreator.create(apiExitCallInfo.getExitCallCount()));
+              addEnrichedAttribute(
+                  event,
+                  EnrichedSpanConstants.API_CALLEE_NAME_COUNT_ATTRIBUTE,
+                  AttributeValue.newBuilder()
+                      .setValueMap(apiExitCallInfo.getCalleeNameToCallCount())
+                      .build());
+            });
+  }
+
+  /**
+   * An api_node represents an api call in the trace It consists of api_entry span and multiple
+   * api_exit and internal spans
+   *
+   * <p>This method computes the count & breakup of exit calls for any given api (identified by
+   * api_entry_span) This count is a composition of 2 things 1. link between api_exit_span in
+   * api_node to api_entry_span (child of api_exit_span) in another api_node 2. exit calls to
+   * backend from api_exit_span in api_node
+   */
+  Map<ByteBuffer, ApiExitCallInfo> computeApiExitCallCount(StructuredTrace trace) {
+    ApiTraceGraph apiTraceGraph = new ApiTraceGraph(trace);
+    // event -> api exit call count for the corresponding api_node
+    Map<ByteBuffer, ApiExitCallInfo> eventToExitInfo = Maps.newHashMap();
+
+    for (ApiNode<Event> apiNode : apiTraceGraph.getApiNodeList()) {
+      List<Event> backendExitEvents =
+          apiTraceGraph.getExitBoundaryEventsWithNoOutboundEdgeForApiNode(apiNode);
+      List<ApiNodeEventEdge> edges = apiTraceGraph.getOutboundEdgesForApiNode(apiNode);
+      int totalExitCallCount = backendExitEvents.size() + edges.size();
+      ApiExitCallInfo apiExitCallInfo = new ApiExitCallInfo().withExitCallCount(totalExitCallCount);
+      edges.forEach(edge -> apiExitCallInfo.handleApiNodeEdge(trace, edge));
+      backendExitEvents.forEach(apiExitCallInfo::handleBackend);
+      apiNode.getEvents().forEach(e -> eventToExitInfo.put(e.getEventId(), apiExitCallInfo));
+    }
+    return eventToExitInfo;
+  }
+
+  static class ApiExitCallInfo {
```

Comment:
nit: Can we move this to its own file?

Reply:
I feel it is internal to this class, and better suited here?